Task manager repair tasks

Deexie opened the pull request:
Integration of task manager with repair module. Allows tracking repair 
operations with task manager.

Currently we can distinguish four main parts of the PR, each of which
concerns the consistent purpose. Some of them may be overly
fragmented in order to easily drop part of the solution if it is divergent
to what is expected. The mentioned parts are:
1. Consisting of `repair: create repair module and task` and
    `add task_manager::module to repair_service`. Introduces task and
    module specific to repair operation and adds it to repair service.
2. Consisting of `repair: change UUID to task_id` and `change UUID to
    task_id further part`. Changes `utils::UUID`to `task_id`. The first commit 
    applies the change to the "high level" code which is strongly connected
    with task manager tasks. The latter goes a little bit further. I am not sure
    whether `task_id` is something specific enough to be used through all
    the way. If we apply tasks to many modules, then it will probably lose its
    main advantage - the distinction of different types of uuids.
3. Consisting of `repair: create root task for user requested repair` and
    `repair: root task for data sync`. Creates top level tasks for repair
    operation (the ones spanning over the whole tasks and kept on shard 0).
    Their scope needs to be discused since the possible solutions are
    connected with some significant trade-offs.
4.  Consisting of `repair: move next_repair_command to repair_module`, 
    `repair: move repairs maps and connected methods to repair module`, 
    and `repair: move the rest of tracer to repair module`. Transfers methods
    and attributes of `tracer` to `repair_module`. The right balance between
    redundancy and the semantics of some repair operations needs to be
    found. Strongly relies on the decisions taken in the third part.

Per shard and per range tasks are to be done.

Deexie commented:
@bhalevy @amnonh @denesb review please

Deexie commented on repair/repair_task.hh:
```diff
@@ -56,7 +56,34 @@ protected:
 };
 
 class repair_module : public task_manager::module {
+    // TODO Deexie:
```
Excuse me this part, but many things happen here and I wanted to keep everything that needs to be thought through in one place as well as give you an easy insight to all that concerns.

bhalevy commented on db/system_keyspace.cc:
```diff
@@ -2903,14 +2903,14 @@ future<> system_keyspace::get_compaction_history(compaction_history_consumer&& f
 future<> system_keyspace::update_repair_history(repair_history_entry entry) {
     sstring req = format("INSERT INTO system.{} (table_uuid, repair_time, repair_uuid, keyspace_name, table_name, range_start, range_end) VALUES (?, ?, ?, ?, ?, ?, ?)", REPAIR_HISTORY);
-    co_await execute_cql(req, entry.table_uuid.uuid(), entry.ts, entry.id, entry.ks, entry.cf, entry.range_start, entry.range_end).discard_result();
+    co_await execute_cql(req, entry.table_uuid.uuid(), entry.ts, entry.id.uuid(), entry.ks, entry.cf, entry.range_start, entry.range_end).discard_result();
```
why drop this patch?
ok to keep it as standalone, ir squash with previous patch.

denesb replied:
If the previous patch doesn't compile without this one, they have to be squashed.

bhalevy reviewed (commented):
I still need to review the root task patch.
submitting some minor comments up to there...

Deexie commented:
- `#include "repair_task.hh"` -> `#include "repair/repair_task.hh"`
- deleted `repair_module_ptr` typedef
- `task_manager_module()` returns reference to pointer

denesb reviewed (commented):
Commits need a lot more text to explain what they are doing and why.

@asias please review too.

bhalevy reviewed (changes requested):
Looks good over all.
See some minor comments.

@asias please review too

Deexie commented:
- changed type of operation for user defined repairs
- squashed commits connected with `task_id`
- added `get_repair_uniq_id()` method
- in `do_repair_start()` sequence number isn't got from the task status
- changed commit messages

bhalevy commented on repair/repair.cc:
```diff
@@ -8,6 +8,7 @@
 
 #include "repair.hh"
 #include "repair/row_level.hh"
+#include "repair_task.hh"
```
nit: `repair/` path prefix

bhalevy replied:
Also, regarding patch subject, the user requested task is derived from the root task, which is only internal to the repair tasks implementation, so I don't think it "deserves" to be mentioned in the subject and commit log.

It would be clearer to describe this as e.g. `repair: create task for user requested repair` or `repair: add user_requested_repair_task_impl`

bhalevy commented on repair/repair_task.hh:
```diff
@@ -23,6 +23,30 @@ public:
+    root_repair_task_impl(task_manager::module_ptr module, task_id id, unsigned sequence_number = 0, std::string keyspace = "", std::string table = "", std::string type = "", std::string entity = "")
+        : repair_task_impl(module, id, sequence_number, std::move(keyspace), std::move(table), std::move(type), std::move(entity)) {}
+protected:
+    repair_uniq_id get_repair_uniq_id() {
```
const noexcept

bhalevy commented on repair/repair.cc:
```diff
@@ -1289,17 +1289,16 @@ future<> repair_service::sync_data_using_repair(
-        streaming::stream_reason reason,
-        std::optional<tasks::task_id> ops_uuid) {
-    seastar::sharded<replica::database>& db = get_db();
+future<> tasks::data_sync_repair_task_impl::run() {
```
Ditto patch subject / commitlog.

It would be clearer to describe this as e.g. `repair: create task for data sync` or `repair: add data_sync_repair_task_impl`

bhalevy commented on repair/repair_task.hh:
```diff
@@ -23,6 +23,30 @@ public:
+    user_requested_repair_task_impl(task_manager::module_ptr module, task_id id, unsigned sequence_number = 0, std::string keyspace = "", std::string table = "", std::string type = "", std::string entity = "", std::unordered_map<sstring, sstring> options_map = {})
+        : root_repair_task_impl(module, id, sequence_number, std::move(keyspace), std::move(table), std::move(type), std::move(entity)), _options_map(std::move(options_map)) {}
+protected:
+    future<> run() override;
```
1. Let's add a TODO for implementing abort for user-initiated repair.
2. How are the repair tasks aborted today? Should / can we integrate that with the task_manager abort source?

Deexie replied:
2. When repair tasks are aborted, the `abort` flag is set to `true`. The flag's value is checked before some operation and an exception is thrown if the flag was set.
Abort source of `task_manager` is the one that is activated on shutdown. It is something different from that abort, but there is a analogical `shutdown` flag in repairs and I believe this one may be integrated. I will take a closer look on that.

bhalevy replied:
as we discussed verbally, let's stick to the current abort design while keeping the repair tasks non-abortable from the task_manager api.
Then later we can transition to using an abort_source per task so we can abort individual tasks, and on top of subscribe to the higher level abort_source and when call on abort, individually abort all outstanding tasks and wait on them to resolve and unregister.

bhalevy commented on repair/repair_task.hh:
```diff
@@ -47,6 +47,20 @@ protected:
+    streaming::stream_reason _reason;
+    std::optional<tasks::task_id> _ops_uuid;
+public:
+    // TODO Deexie null task id instead of optional
```
1. when is it set to null?
2. should it be a member of `root_repair_task_impl` so to be provided as a parent task_id?

Deexie replied:
1. It is set to null for each data sync repair except for `removenode`. The only reason I can see behind that optional is a condition:
https://github.com/scylladb/scylladb/blob/6b180bf2c7830a5930ffcc87a9e4cf3a25db9092/repair/repair.cc#L450-L453
Where it is clearer that we cast it to bool on lhs and then compare the value on rhs than if we had: 
```

bhalevy commented on repair/repair_task.hh:
```diff
@@ -47,6 +47,20 @@ protected:
+    data_sync_repair_task_impl(task_manager::module_ptr module, task_id id, unsigned sequence_number = 0, std::string keyspace = "", std::string table = "", std::string type = "", std::string entity = "", dht::token_range_vector ranges = {}, std::unordered_map<dht::token_range, repair_neighbors> neighbors = {}, streaming::stream_reason reason = streaming::stream_reason::unspecified, std::optional ...
+        : root_repair_task_impl(module, id, sequence_number, std::move(keyspace), std::move(table), std::move(type), std::move(entity)), _ranges(std::move(ranges)), _neighbors(std::move(neighbors)), _reason(reason), _ops_uuid(ops_uuid) {}
+protected:
+    future<> run() override;
```
ditto TODO for abort, like user_requested_repair_task_impl

bhalevy commented on repair/repair.cc:
```diff
@@ -343,32 +343,35 @@ void tracker::done(repair_uniq_id id, bool succeeded) {
-    if (id >= _next_repair_command) {
+
+repair_status tasks::repair_module::get(int id) const {
+    if (id > _sequence_number) {
```
why `>` and not `>=`?

Deexie replied:
`_next_repair_command` is a number of next repair command while `_sequence_number` keeps the value of the last number assigned to some task.
I haven't changed that in `tasks::repair_module::repair_await_completion`, though. Will do.

bhalevy commented on repair/repair.cc:
```diff
@@ -343,32 +343,35 @@ void tracker::done(repair_uniq_id id, bool succeeded) {
-    if (it == _status.end()) {
+    auto& tracker = _rs.repair_tracker();
+    auto it = tracker._status.find(id);
+    if (it == tracker._status.end()) {
```
There seems to be duplication still between the repair_tracker status tracking and the repair tasks module.
Let's add a TODO to track the tasks entirely by the task module, and maybe even get rid of the repair tracker entirely, eventually.

bhalevy commented on repair/repair_task.hh:
```diff
@@ -62,7 +62,34 @@ protected:
+    // - where tracer::run starts. This way _status, _pending_repairs, and _aborted_pending_repairs could freely cooperate with task_manager::module::task.
+    // The problem with logs order appears (some data is logged in the background).
+    //
+    // The order of shutting down. Can we use common gate? Can we get rid of shutdown methods?
```
I think we should use a the module gate, along with integrating with the task manager abort source.

bhalevy commented on repair/repair_task.hh:
```diff
@@ -62,7 +62,34 @@ protected:
+    //  even the ones that were terminated in preliminary correctness check. This way we will also avoid inconvenience caused by the task id being logged before
+    //  the task is actually created.
+    // - where tracer::run starts. This way _status, _pending_repairs, and _aborted_pending_repairs could freely cooperate with task_manager::module::task.
+    // The problem with logs order appears (some data is logged in the background).
```
@asias what do you think?

bhalevy commented on repair/repair.cc:
```diff
@@ -58,9 +58,8 @@ void node_ops_info::check_abort() {
 
-node_ops_metrics::node_ops_metrics(tracker& tracker, repair_module_ptr module)
-    : _tracker(tracker)
-    , _module(module)
```
nit: s/tracer/tracker/g in patch subject and commit log

bhalevy commented on repair/repair_task.hh:
```diff
@@ -82,6 +89,10 @@ class repair_module : public task_manager::module {
     // The order of shutting down. Can we use common gate? Can we get rid of shutdown methods?
 private:
+    // Used to allow shutting down repairs in progress, and waiting for them.
+    seastar::gate _repair_gate;
```
why not use the base class `_gate`?

Deexie replied:
I haven't done that because I did not understand why repair shutdown is divided into three part and I didn't wanted to mess it up. I think I finally managed to get it but I need your confirmation.

Repair shutdown is divided into 3 parts (`main.cc`):
- `repair_service::stop()`
- repair api is unregistered
- `repair_service::shutdown()`
which happen to be called bottom to top.

This is done because we want to stop all the repairs, while api is still available. Then, unregister the api. And then, close whole `repair_service`.

If this is true, then I should surely move `module::stop()` from `repair_service::stop()` to `repair_service::shutdown()`.
And then we could freely use the `_gate` and maybe even delete `with_gate` from:
https://github.com/scylladb/scylladb/blob/6b180bf2c7830a5930ffcc87a9e4cf3a25db9092/repair/repair.cc#L474-L485
Since it is already protected by `task_manager::task` gate.

One thing that comes to mind here is if task manager api should have been unregistered. Some of the already existing services are unregistered, some are not. I do not see the pattern of choice.

Also I think that the shutdown should happen in the following order:
- stop tasks
- unregister services' apis
- (unregister task manager api)
- close task_manager::modules
- close task_manager
- close services
Thus, be called in `main` in reversed order.

What do you think?

bhalevy replied:
> I haven't done that because I did not understand why repair shutdown is divided into three parts and I didn't wanted to mess it up. I think I finally managed to get it but I need your confirmation.
> 
> Repair shutdown is divided into 3 parts (`main.cc`):
> 
> * `repair_service::stop()`
> * repair api is unregistered
> * `repair_service::shutdown()`
>   which happen to be called bottom to top.
> 
> This is done because we want to stop all the repairs, while api is still available. Then, unregister the api. And then, close whole `repair_service`.
> 
> If this is true, then I should surely move `module::stop()` from `repair_service::stop()` to `repair_service::shutdown()`.

By convention, shutdown should just signal whatever is running to abort, here, it could happen using the task manager abort_source.
Then, stop() waits for everything to actually stop before destroying the object.

So the repair modude stop can remain being called from repair service stop, given the abort is requested at shutdown time and is being consulted on the the task run path (not to start any new tasks, checked in loops, e.g. per each token range), and used for sleep_abirtable and for waiting e.g. sem.wait or on async condition variables.

> And then we could freely use the `_gate` and maybe even delete `with_gate` from:
> 
> https://github.com/scylladb/scylladb/blob/6b180bf2c7830a5930ffcc87a9e4cf3a25db9092/repair/repair.cc#L474-L485
> 
> 
> Since it is already protected by `task_manager::task` gate.
> One thing that comes to mind here is if task manager api should have been unregistered. Some of the already existing services are unregistered, some are not. I do not see the pattern of choice.
> 
> Also I think that the shutdown should happen in the following order:
> 
> * stop tasks

that would be requesting abort via the abort source. At this point no new tasks should start even while the api is still registered.  They could return an appropriate error at this state.

> * unregister services' apis
> * (unregister task manager api)

that will happen when the task manager service is stopped.

> * close task_manager::modules
> * close task_manager
> * close services
>   Thus, be set in `main` in reversed order.

right

> 
> What do you think?

Cc @xemul

asias commented:
Some comments below:


0) Let's add more background and motivation to the cover letter.
```
Why do we want to introduce the task module for repair?
What are the benefits?
Any impact for the user to use repair?
```

1) The name root_repair_task_impl is confusing. 
It is root in what sense. Can we get rid of it so we can have less indirection?

2) Who is going to keep the task object alive after do_repair_start returns?
```
int repair_service::do_repair_start(sstring keyspace, std::unordered_map<sstring, sstring> options_map) {
        auto sequence_number = _task_manager_module->new_sequence_number();
        auto it = options_map.find(repair_options::COLUMNFAMILIES_KEY);
        sstring table = it == options_map.end() ? "" : it->second;
        auto task_impl_ptr = std::make_unique<tasks::user_requested_repair_task_impl>(_task_manager_module, tasks::task_id::create_random_id(), sequence_number, std::move(keyspace), std::move(table), format("{}", streaming::stream_
reason::repair), "", std::move(options_map));
        auto task = make_lw_shared<tasks::task_manager::task>(std::move(task_impl_ptr));
        task->start();
        return sequence_number;
}
```
3) In this patch, task_id is used to present repair uuid / node operation uuid
and the repair history.  The point of using task_id instead of uuid is to make
the type same. Using task_id for everything is not helping the goal. It is not
making it worse though. I think task_id is not concrete enough. We can call
anything a task. Why not use the real type name, e.g., repair_id and
node_ops_id instead?

4) shared_ptr<tasks::repair_module> _task_manager_module;
 shared_ptr<tasks::repair_module>& task_manager_module() noexcept

  Can we just call it _repair_module and repair_module() ? So we have less name to worry about and make it more consistent.

raphaelsc commented on types.hh:
```diff
@@ -33,6 +33,7 @@
 #include "utils/bit_cast.hh"
 #include "utils/chunked_vector.hh"
+#include "tasks/task_manager.hh"
 
```
types.hh is a heavily used header, so I recommend pushing tasks::task_id into tasks/types.hh.

Deexie commented:
- added "repair/" to includes
- deleted "root" from commits' messages
- added TODOs
- `>=` -> `>` in `tasks::repair_module::repair_await_completion`
- updated cover letter
- deleted `root_repair_task_impl` class
- `task_manager_module` -> `repair_module`
- moved `task_id` to `tasks/types.hh`
- repair `used_id` and `node_ops_id`
- deleted repair_gate
- used abort source instead of shutdown flag

Deexie commented:
>     2. Who is going to keep the task object alive after do_repair_start returns?
> 
> 
> ```
> int repair_service::do_repair_start(sstring keyspace, std::unordered_map<sstring, sstring> options_map) {
>         auto sequence_number = _task_manager_module->new_sequence_number();
>         auto it = options_map.find(repair_options::COLUMNFAMILIES_KEY);
>         sstring table = it == options_map.end() ? "" : it->second;
>         auto task_impl_ptr = std::make_unique<tasks::user_requested_repair_task_impl>(_task_manager_module, tasks::task_id::create_random_id(), sequence_number, std::move(keyspace), std::move(table), format("{}", streaming::stream_
> reason::repair), "", std::move(options_map));
>         auto task = make_lw_shared<tasks::task_manager::task>(std::move(task_impl_ptr));
>         task->start();
>         return sequence_number;
> }
> ```

`task_manager::task` registers itself (in `task_manager` and `task_manager::module`) in its constructor.

Deexie commented:
- add range tasks
- `repair_id` -> `task_id`
- gather progress in tasks

Deexie commented on idl/partition_checksum.idl.hh:
```diff
@@ -8,6 +8,7 @@
 
 #include "dht/i_partitioner.hh"
```
Temporary commit before https://github.com/scylladb/scylladb/pull/11673 gets merged.

Deexie commented:
- modify commit messages
- moved `repair_options` to `repair_task.hh`
- change the range over which top level tasks span

bhalevy commented on types.hh:
```diff
@@ -1041,6 +1042,12 @@ shared_ptr<const abstract_type> data_type_for<cql_duration>() {
+inline
+shared_ptr<const abstract_type> data_type_for<tasks::task_id>() {
+    return uuid_type;
+}
```
@Deexie please try the following:
```diff
diff --git a/types.hh b/types.hh
index fc71898dd7..a8f7d5d5f2 100644
--- a/types.hh
+++ b/types.hh
@@ -1042,9 +1042,12 @@ shared_ptr<const abstract_type> data_type_for<cql_duration>() {
     return duration_type;
 }
 
-template <>
+template <typename Tag>
+requires requires (utils::tagged_uuid<Tag> id) {
+    { id.uuid() } -> std::same_as<utils::UUID>;
+}
 inline
-shared_ptr<const abstract_type> data_type_for<tasks::task_id>() {
+shared_ptr<const abstract_type> data_type_for() {
     return uuid_type;
 }
 
```

Deexie replied:
`ld.lld: error: undefined symbol: seastar::shared_ptr<abstract_type const> data_type_for<utils::tagged_uuid<tasks::task_id_tag> >()`

bhalevy replied:
all right, I give up :)
let's leave it the way it is.

bhalevy commented on repair/repair.cc:
```diff
@@ -995,112 +823,14 @@ static future<> repair_ranges(lw_shared_ptr<repair_info> ri) {
-    if (cfs.empty()) {
-        rlogger.info("repair[{}]: completed successfully: no tables to repair", id.uuid);
-        return id.id;
-    }
```
to improve the patch review, can we refactor the code to keep the code in place here
and call it from the new framework, and then maybe as a second step move it there as is?

bhalevy commented on repair/repair.hh:
```diff
@@ -214,15 +218,11 @@ public:
-    int _next_repair_command = 1;
     // Note that there are no "SUCCESSFUL" entries in the "status" map:
-    // Successfully-finished repairs are those with id < _next_repair_command
+    // Successfully-finished repairs are those with id < repair_module::_sequence_number
```
nit: s/</<=/

bhalevy commented on repair/repair_task.hh:
```diff
@@ -250,8 +250,29 @@ protected:
 class repair_module : public task_manager::module {
+    // TODO Deexie:
+    // _status - the difference between _status and modules maps that makes us keep this map is that the information about tasks that failed
+    // is kept forever (successful one are deleted after they are finished). Thus, we can keep the map as it is or keep a set of sequence numbers of failed
```
can we erase the status after we returned the error to the user? and/or after the task ttl expires?

Deexie replied:
This would change the existing repair api. Currently we can get the status of all tasks that happened in the past - the ones that failed are kept in `_status` forever.

bhalevy commented on repair/repair.cc:
```diff
@@ -796,17 +796,22 @@ static future<> do_repair_ranges(lw_shared_ptr<repair_info> ri) {
+    }
+
+    _ri->rs.repair_module()->add_repair_info(_ri->id.id, _ri);
+    return do_repair_ranges(_ri).then([ri = _ri] {
```
_ri is kept alive as a member of the task.
I suggest to get a reference to it in the continuation body rather than capturing the shared ptr and capture `this` instead.

bhalevy commented on repair/repair.cc:
```diff
@@ -796,17 +796,22 @@ static future<> do_repair_ranges(lw_shared_ptr<repair_info> ri) {
         ri->rs.repair_module()->remove_repair_info(ri->id.id);
         return make_ready_future<>();
-    }).handle_exception([ri] (std::exception_ptr eptr) {
+    }).handle_exception([ri = _ri] (std::exception_ptr eptr) {
```
ditto here

bhalevy commented on repair/repair.cc:
```diff
@@ -914,12 +919,21 @@ future<> tasks::user_requested_repair_task_impl::run() {
+                    ex = std::current_exception();
+                }
+                auto task_impl_ptr = std::make_unique<tasks::shard_repair_task_impl>(local_repair._repair_module, task_id::create_null_id(), 0, keyspace, "", format("{}", streaming::stream_reason::repair), "", task_id, ri, ex);
+                auto task = co_await start_repair_task(std::move(task_impl_ptr), local_repair._repair_module, task_manager::parent_data{task_id, this_shard_id()});
```
don't we need to get the parent shard in the calling shard, ahead of the invoke_on contiuation?
maybe we can keep it in the repair id structure so to have the sequence number and parent_data (uuid and owning shard)

bhalevy commented on repair/repair.cc:
```diff
@@ -1127,16 +1141,25 @@ future<> tasks::data_sync_repair_task_impl::run() {
+                    ex = std::current_exception();
+                }
+                auto task_impl_ptr = std::make_unique<tasks::shard_repair_task_impl>(local_repair._repair_module, task_id::create_null_id(), 0, keyspace, "", format("{}", streaming::stream_reason::repair), "", task_id, ri, ex);
+                auto task = co_await start_repair_task(std::move(task_impl_ptr), local_repair._repair_module, task_manager::parent_data{task_id, this_shard_id()});
```
ditto parent shard id

bhalevy replied:
how is it tested?  do we have any in-tree tests for repair?

bhalevy commented on repair/repair.cc:
```diff
@@ -744,7 +744,11 @@ static future<tasks::task_manager::task_ptr> start_repair_task(tasks::task_manag
-static future<> do_repair_ranges(lw_shared_ptr<repair_info> ri) {
+future<> tasks::range_repair_task_impl::run() {
+    return _ri->repair_range(_range, _table_id);
+}
```
nit: in the commit log message s/spaning/spanning/

bhalevy commented on repair/repair.cc:
```diff
@@ -744,7 +744,11 @@ static future<tasks::task_manager::task_ptr> start_repair_task(tasks::task_manag
+    return _ri->repair_range(_range, _table_id);
+}
+
+static future<> do_repair_ranges(lw_shared_ptr<repair_info> ri, tasks::task_id task_id) {
```
why isn't this a method of the upper-level task?

bhalevy commented on repair/repair.cc:
```diff
@@ -753,32 +757,36 @@ static future<> do_repair_ranges(lw_shared_ptr<repair_info> ri) {
+            auto ri = tmp_ri;
+            auto task_impl_ptr = std::make_unique<tasks::range_repair_task_impl>(ri->rs.repair_module(), tasks::task_id::create_null_id(), 0, ri->keyspace, "", format("{}", ri->reason), "", task_id, ri, range, table_id);
+            auto id = co_await ri->rs.repair_module()->make_task(this_shard_id(), std::move(task_impl_ptr), tasks::task_manager::parent_data{task_id, this_shard_id()});
+            auto task = ri->rs.repair_module()->get_tasks()[id];
```
I wonder if make_task should return the id or the task foreign shared ptr then
you did the same for other tasks as well and if this is the common use case maybe we should change `make_task` to return what its callers need.

bhalevy replied:
there should be no problem accessing the task->id() on the same shard and even on a foreign shard.

Deexie replied:
I think we should make `make_task` return `task_ptr` (and not `foreign_task_ptr`), since it gets a pointer to task impl as an argument and we should rather allow that only on target shard than play with foreign's.

The old `make_task` should stay as it is though I believe.

bhalevy replied:
makes sense

bhalevy commented on repair/repair_task.hh:
```diff
@@ -259,6 +278,10 @@ public:
         , _ex(ex)
     {}
+
+    future<task_manager::task::progress> gather_child_progress(unsigned i) const override {
```
should be virtual override

bhalevy replied:
that said, submit_to the same shard costs very little, so I'm not sure virtualizing this method makes sense.
It doesn't save much in this case.
Do we have a need for a radically different way of getting the child progress?

bhalevy commented on repair/repair_task.hh:
```diff
@@ -275,6 +298,14 @@ public:
         , _table_id(table_id)
     {}
+
+    future<task_manager::task::progress> get_progress() const override {
```
nit: virtual override

Deexie commented on tasks/task_manager.hh:
```diff
@@ -346,9 +356,29 @@ public:
+
+            auto task = co_await _tm.container().invoke_on(shard, [task_impl_ptr = std::move(task_impl_ptr), sequence_number] (task_manager& tm) mutable {
+                auto task = make_lw_shared<task_manager::task>(std::move(task_impl_ptr));
+                task->get_status().sequence_number = sequence_number == 0 ?  task->get_module()->new_sequence_number() : sequence_number;
```
I have no clue what I tried to do here, but this function does not make any sense on many layers...

I am rewriting it from the very beginning. The function would be expected to be called from the child's shard and will return task_ptr. The old `make_task` will still return `task_id`, though, since it can be called from any shard.

Deexie commented:
- set repair progress units to "ranges"
- modified `get_parent()` to be regular continuation
- rewrote `make_task`
- changed order of methods definition to make diffs easily comparable
- captured this instead of its members in lambdas
- added shard id to `repair_uniq_id`
- `do_repair_ranges` and `repair_ranges` were made a method of `repair_task_impl`
- `gather_child_progress` is done the same way for each task type

bhalevy commented:
@denesb, @asias  can you please review the latest version?
How far are we for this PR to be mergeable?

bhalevy commented:
@Deexie please move the PR out of Draft mode when you think it's ready for final review and merging, i.e
 contains the minimum required functionality and addressed all review comments.

Please also summarize the targetted testing coverage we have - in tree and in dtests. See what dtests cover this in CI, others should be run using the BYO job on Jenkins

denesb commented on repair/repair.cc:
```diff
@@ -1125,16 +1139,25 @@ future<> tasks::data_sync_repair_task_impl::run() {
+                lw_shared_ptr<repair_info> ri;
+                std::exception_ptr ex;
+                auto task_id = tasks::task_id{id.uuid().uuid()};
+                try {
```
You can avoid such re-indent churn by adding the `try {` at a half-indent so code doesn't have to be re-indented. Then re-indent in a separate commit.

denesb commented on repair/repair_task.hh:
```diff
@@ -252,6 +251,20 @@ protected:
+    lw_shared_ptr<repair_info> _ri;
+    std::exception_ptr _ex;
+public:
+    shard_repair_task_impl(task_manager::module_ptr module, task_id id, unsigned sequence_number = 0, std::string keyspace = "", std::string table = "", std::string type = "", std::string entity = "", task_id parent_id = task_id::create_null_id(), lw_shared_ptr<repair_info> ri = nullptr, std::exception_ptr ex = nullptr)
```
Why so many default parameters? Is there any user relying on them? Is an object constructed with them valid?

denesb commented on repair/repair.cc:
```diff
@@ -1125,16 +1139,25 @@ future<> tasks::data_sync_repair_task_impl::run() {
+                    ex = std::current_exception();
+                }
+                auto task_impl_ptr = std::make_unique<tasks::shard_repair_task_impl>(local_repair._repair_module, task_id::create_random_id(), 0, keyspace, "", format("{}", streaming::stream_reason::repair), "", task_id, ri, ex);
+                auto task = co_await start_repair_task(std::move(task_impl_ptr), local_repair._repair_module, parent_data);
```
We don't need to do `task->start()` here?

Deexie replied:
`start_repair_task` does it.

denesb replied:
I see.

denesb reviewed (changes requested):
Looks good to me on a high level. I had some questions and minor comments, nothing major.
The types of tasks and their structure makes sense.
I wish we could also track more metrics on a per-task basis, like these here https://github.com/scylladb/scylladb/blob/fcf0628bc5d03f13d75f12e1e639747a178821bf/repair/row_level.cc#L168-L178. Can be done as a follow-up.
It would be nice if the cover-letter contained an example of how all this looks like in practice: setup a small cluster (can be just 2 nodes on your laptop, but each should have at least 2 shards), write some data, start a repair and then query the task metadata and copy the result into the cover-letter.

The series is quite hard to review, commits often do too much and mix multiple things. In general:
* Commits should never mix code movement with anything else. The intent is that when a reviewer sees that a commit is just moving around code, they don't have to check the changes, as they can be sure no changes are done besides cut+paste.
* Commits should not mix changes with code re-indent. It is preferable to leave the indentation broken, then have a later commit fix it. This can be even a single commit at the end of the series fixing all broken indentation. This is fine as long as that commit just does re-indenting.
* When changes you do require a lot of churn, e.g. you have to rename a variable from `_ri` to `ri`, it is preferable to do this change beforehand, even if alone this is artificial, e.g. `auto& ri = _ri;` (and update all usages in the method). Such commits are good because they are a no-brainer to the reviewer. While if they are mixed with other changes, these simple renames also have to be examined to see if there is something more going on.
* Coroutinizing some code should also be done as a separate commit.

Deexie commented on repair/row_level.hh:
```diff
@@ -88,6 +90,7 @@ class repair_service : public seastar::peering_sharded_service<repair_service> {
     sharded<db::system_distributed_keyspace>& _sys_dist_ks;
     sharded<db::system_keyspace>& _sys_ks;
     sharded<db::view::view_update_generator>& _view_update_generator;
+    shared_ptr<tasks::repair_module> _repair_module;
```
Task manager keeps a map of shared pointers to modules. It is registered there in `repair_service` constructor.

denesb replied:
I see. Makes sense then.

Deexie commented on repair/repair.cc:
```diff
@@ -747,185 +748,11 @@ sstring repair_stats::get_stats() {
-    }
-};
-
+static future<tasks::task_manager::task_ptr> start_repair_task(tasks::task_manager::task::task_impl_ptr task_impl_ptr, shared_ptr<tasks::repair_module> module, tasks::parent_data pd = {}) {
```
Task registers itself in its ctor, which is called in `make_task`. Then it is kept by task manager and repair module.

Deexie commented on repair/repair.cc:
```diff
@@ -467,25 +459,23 @@ float tracker::report_progress(streaming::stream_reason reason) {
+        done(id, true);
+    }).handle_exception([this, id] (std::exception_ptr ep) {
+        done(id, false);
+        return make_exception_future(std::move(ep));
```
The job of this gate is performed by module `_gate` when the task is run.

Deexie commented:
- deleted `tasks` namespace from `repair/repair_task.hh`
- `get_repair_module` in `repair_service` returns reference instead of shared pointer
- `parent_data` -> `task_info`
- deleted shard from repair_info
- deleted default params from repair tasks
- split commits into smaller parts
- added commit that changes indentation
- rebased on master
- changed `do_repair_ranges` so that it saves captures
- added test

asias commented:
> If repair is a module of task manager then users can easily check which
repairs and node operations are currently running or finished, what is
their progress and status. They will also be able to abort a chosen
repair operation.

It is not clear how users can use task manager stuff to see the status and progress. Can you add concrete examples in the commit message?

asias commented on repair/repair.cc:
```diff
@@ -915,12 +921,25 @@ int repair_service::do_repair_start(sstring keyspace, std::unordered_map<sstring
-            cfs = std::move(cfs), ranges = std::move(ranges), options = std::move(options), ignore_nodes = std::move(ignore_nodes)] () mutable {
+    auto it = options_map.find(repair_options::COLUMNFAMILIES_KEY);
+    sstring table = it == options_map.end() ? "" : it->second;
+    auto task_impl_ptr = std::make_unique<user_requested_repair_task_impl>(_repair_module, id, std::move(keyspace), std::move(table), format("{}", streaming::stream_reason::repair), "", std::move(cfs), std::move(ranges), std::move(options), std::move(ignore_nodes));
```
The table is a incorrect abstraction for user_requested_repair_task_impl.

A user task should be able to repair multiple tables. Instead of a single table.

asias commented on repair/repair.cc:
```diff
@@ -915,12 +921,25 @@ int repair_service::do_repair_start(sstring keyspace, std::unordered_map<sstring
-    // Do it in the background.
-    (void)repair_tracker().run(id, [this, &db, id, keyspace = std::move(keyspace),
-            cfs = std::move(cfs), ranges = std::move(ranges), options = std::move(options), ignore_nodes = std::move(ignore_nodes)] () mutable {
+    auto it = options_map.find(repair_options::COLUMNFAMILIES_KEY);
```
No need to query the options_map here. 

Just a fines above we already have:

```
    std::vector<sstring> cfs =         
        options.column_families.size() ? options.column_families : list_column_families(db.local(), keyspace);
    if (cfs.empty()) {
        rlogger.info("repair[{}]: completed successfully: no tables to repair", id.uuid());
        return make_ready_future<int>(id.id);
    }

```

asias commented:
Unrelated change in this PR:

$ git diff f673d0abbe752c1f7b48130e638c737c5f16c8bf --stat|grep abseil
 abseil                             |   2 +-

asias commented on repair/repair.cc:
```diff
@@ -756,9 +756,14 @@ future<> repair_task_impl::do_repair_ranges(lw_shared_ptr<repair_info> ri) {
-                return ri->repair_range(range, table_id).then([ri] {
+        co_await coroutine::parallel_for_each(ri->ranges, [ri, table_id, parent_data] (auto&& range) -> future<> {
+            auto tmp_ri = ri;
+            auto task_impl_ptr = std::make_unique<range_repair_task_impl>(ri->rs.get_repair_module().shared_from_this(), tasks::task_id::create_random_id(), 0, ri->keyspace, "", format("{}", ri->reason), "", parent_data.id, ri, range, table_id);
```
Why does the table here is pass as empty string?

What is the the empty  "" after reason?

asias replied:
why is the seq number always zero?

Deexie replied:
Sequence number for child tasks (shard and range ones) are inherited from parents. There is no need to pass that to the ctor, I will delete it.
And the table would be set to `table_name`.

asias commented on repair/repair.cc:
```diff
@@ -756,9 +756,14 @@ future<> repair_task_impl::do_repair_ranges(lw_shared_ptr<repair_info> ri) {
-            return with_semaphore(ri->rs.get_repair_module().range_parallelism_semaphore(), 1, [ri, &range, table_id] {
-                return ri->repair_range(range, table_id).then([ri] {
+        co_await coroutine::parallel_for_each(ri->ranges, [ri, table_id, parent_data] (auto&& range) -> future<> {
+            auto tmp_ri = ri;
```
Do we really need a tmp_ri?

asias commented on repair/repair.cc:
```diff
@@ -756,9 +756,14 @@ future<> repair_task_impl::do_repair_ranges(lw_shared_ptr<repair_info> ri) {
+        co_await coroutine::parallel_for_each(ri->ranges, [ri, table_id, parent_data] (auto&& range) -> future<> {
+            auto tmp_ri = ri;
+            auto task_impl_ptr = std::make_unique<range_repair_task_impl>(ri->rs.get_repair_module().shared_from_this(), tasks::task_id::create_random_id(), 0, ri->keyspace, "", format("{}", ri->reason), "", parent_data.id, ri, range, table_id);
+            auto task = co_await ri->rs.get_repair_module().make_task(std::move(task_impl_ptr), parent_data);
```
```
    range_repair_task_impl(tasks::task_manager::module_ptr module, tasks::task_id id, unsigned sequence_number, std::string keyspace, std::string table, std::string type, std::string entity, tasks::task_id parent_id, lw_shared_ptr<repair_info> ri, dht::token_range range, table_id table_id)
        : repair_task_impl(module, id, sequence_number, std::move(keyspace), std::move(table), std::move(type), std::move(entity), parent_id)
        , _ri(ri)
        , _range(range)
        , _table_id(table_id)
    {}

    virtual future<tasks::task_manager::task::progress> get_progress() const override {
        auto state = _status.state;
        co_return tasks::task_manager::task::progress{
            .completed = state == tasks::task_manager::task_state::done || state == tasks::task_manager::task_state::failed,
            .total = 1.0
        };
    }
```

```
        co_await coroutine::parallel_for_each(ri->ranges, [ri, table_id, parent_data] (auto&& range) -> future<> {
            auto tmp_ri = ri;
            auto task_impl_ptr = std::make_unique<range_repair_task_impl>(ri->rs.get_repair_module().shared_from_this(), tasks::task_id::create_random_id(), 0, ri->keyspace, "", format("{}", ri->reason), "", parent_data.id, ri, range, table_id);
```

I think the way to calculate  the progress is not correct here.

so, each range_repair_task_impl crreated contribute 1 to total range. The correctness of the progress depends on if the 

co_await coroutine::parallel_for_each has finished creating all the task ranges. 

Also, we have

```
future<> repair_task_impl::do_repair_ranges(lw_shared_ptr<repair_info> ri) {
    auto parent_data = get_repair_uniq_id().task_data;
    // Repair tables in the keyspace one after another
    assert(ri->table_names().size() == ri->table_ids.size());
    for (int idx = 0; idx < ri->table_ids.size(); idx++) {
        auto table_id = ri->table_ids[idx];
        auto table_name = ri->table_names()[idx];

```

we process the ranges table by table, so the "total"  is always wrong until the last table is processed.

asias commented on repair/repair.cc:
```diff
@@ -743,11 +743,11 @@ static future<tasks::task_manager::task_ptr> start_repair_task(tasks::task_manag
 future<> range_repair_task_impl::run() {
-    // TODO: implement
-    return make_ready_future<>();
+    return _ri->repair_range(_range, _table_id);
```
To be honest, I am not a big fan of adding tons of foo_task_impl. They are just too many of them which makes it is less useful to for user to track them. 

We  have shard_repair_task_impl and range_repair_task_impl. 
Say we have a repair task with 20 tables and 1024 ranges and 32 shards. 

We are going to create

20 * 1024  * 32 shard_repair_task_impl

and 

20 * 1024  range_repair_task_impl.

There is no way for a user to track this amount of tasks.

bhalevy replied:
Maybe we can add an attribute to the task class to show/hide it in the tasks list.
I think it could be helpful to use the task_manager::task class for internal purposes even if we need not show it over the api.

bhalevy replied:
something like `using internal = bool_class<struct internal_tag>`

asias commented on repair/repair_task.hh:
```diff
@@ -10,6 +10,190 @@
+
+    // A range is expressed as start_token:end token and multiple ranges can
+    // be given as comma separated ranges(e.g. aaa:bbb,ccc:ddd).
+    static void ranges_opt(dht::token_range_vector& var,
```
Let's not move code from .cc to  .hh.

Deexie commented:
Part of this PR that introduces changes to task manager was moved to: https://github.com/scylladb/scylladb/pull/11821

Deexie commented:
- dropped commit moving `repair_options` from `repair.cc` to `repair_task.hh`
- changed attributes of `user_requested_repair_task_impl`
- table field in top level repair tasks' statuses isn't set but it is for `range_repair_task_impl`
- omitted sequence number in child tasks ctor
- added internal flag
- changed order of commits
- fixed the way total progress is counted

Deexie commented:
The PR that introduces changes connected with repair module: #11851

Deexie commented:
The PR that introduces top level tasks and abort operation: #11869

Deexie commented:
The PR that introduces shard repair tasks: #12066

Deexie commented:
The PR that unifies methods of aborting repair tasks: #12085.
